### PR TITLE
signet: omit commitment for some trivial challenges

### PR DESCRIPTION
**Pull request by Sjors**
[BIP325](https://github.com/bitcoin/bips/blob/master/bip-0325.mediawiki) mentions the following rule:

> In the special case where an empty solution is valid (ie scriptSig and scriptWitness are both empty) this additional commitment can optionally be left out. This special case is to allow non-signet-aware block generation code to be used to test a custom signet chain where the challenge is trivially true.

Such a signet can be created using e.g. `-signetchallenge=51` (`OP_TRUE`). However `contrib/signet/miner` won't omit the commitment.

This PR improves the miner by skipping the PSBT for known trivial scripts (just `OP_TRUE` and trivial pushes for now). This prevents it from appending the 4 byte signet header to the witness commitment, as allowed by the above rule.

---

Previously the script would fail with `PSBT signing failed`, making it difficult to mine. This is no longer the case.

### CODE REVIEW

**Comment by Sjors** [comment]
My own goal with this is to test Stratum v2 stuff, e.g. #28983. It's nice to have a test network _with_ difficulty adjustment. Regular testnet difficulty is too high for my humble S9 (let alone a CPU miner) to win blocks.

Using real mining equipment as opposed to a CPU miner is useful, because of the many quirks real devices have (e.g. version bit grinding). Afaik there's no good emulator.

cc @ajtowns, @kallewoof

**Comment by kallewoof** [comment]
This looks reasonable to me, although I'm not super familiar with all the code being changed.

**Review by edilmedeiros — COMMENTED** [review]
Concept ACK

Code looks good, I'm still to test it.

**Comment by Sjors** [comment]
Rebased after #28417.

File: test/functional/tool_signet_miner.py [hunk 1]
```diff
@@ -61,6 +66,14 @@ def run_test(self):
             ], check=True, stderr=subprocess.STDOUT)
         assert_equal(node.getblockcount(), 1)
 
+    def run_test(self):
+        self.log.info("Signet node with trivial challenge (OP_RETURN)")
```
Comment (ajtowns) [inline test/functional/tool_signet_miner.py:70]: `OP_RETURN` ?

File: contrib/signet/miner [hunk 2]
```diff
@@ -322,14 +328,23 @@ class Generate:
...
-            return None
-        block, signet_solution = decode_psbt(psbt_signed["psbt"])
+        block = new_block(tmpl, reward_spk, blocktime=self.mine_time, poolid=self.poolid)
+
+        # BIP325 allows omitting the signet commitment when scriptSig and
+        # scriptWitness are both empty. This is the case for trivial
+        # challenges such as OP_TRUE
+        signet_solution = None
+        signet_spk = tmpl["signet_challenge"]
+        if signet_spk != "51":
```
Comment (ajtowns) [inline contrib/signet/miner:338]: Might be better as something like:

```python
    signet_spk = tmpl["signet_challenge"]
    if trivial_challenge(signet_spk):
        signet_solution = None:
    else:
        ...
```

and

```
def trivial_challenge(spkhex):
    """BIP325 allows ..."""
    if spkhex == "51":
        return True
    return False
```

File: contrib/signet/miner [hunk 3]
```diff
@@ -79,9 +79,10 @@ def decode_psbt(b64psbt):
     return from_binary(CBlock, psbt.g.map[PSBT_SIGNET_BLOCK]), ser_string(scriptSig) + scriptWitness
 
 def finish_block(block, signet_solution, grind_cmd):
-    block.vtx[0].vout[-1].scriptPubKey += CScriptOp.encode_op_pushdata(SIGNET_HEADER + signet_solution)
-    block.vtx[0].rehash()
-    block.hashMerkleRoot = block.calc_merkle_root()
+    if signet_solution is not None:
```
Comment (ajtowns) [inline contrib/signet/miner:82]: nit: I'd probably write this as

```python
if signet_solution is None:
    pass # Don't need to add a signet commitment if there's no signet signature needed
else:
    ...
```

File: contrib/signet/miner [hunk 4]
```diff
@@ -93,10 +94,20 @@ def finish_block(block, signet_solution, grind_cmd):
         block.rehash()
     return block
 
-def generate_psbt(tmpl, reward_spk, *, blocktime=None, poolid=None):
-    signet_spk = tmpl["signet_challenge"]
+def generate_psbt(block, signet_spk):
```
Comment (ajtowns) [inline contrib/signet/miner:97]: Better to move `def generate_psbt(..)` down to its code, rather than moving the code up?

File: contrib/signet/miner [hunk 5]
```diff
@@ -322,14 +328,23 @@ class Generate:
...
+        signet_spk = tmpl["signet_challenge"]
+        if signet_spk != "51":
+            psbt = generate_psbt(block, signet_spk)
+            input_stream = os.linesep.join([psbt, "true", "ALL"]).encode('utf8')
+            psbt_signed = json.loads(bcli("-stdin", "walletprocesspsbt", input=input_stream))
+            if not psbt_signed.get("complete",False):
+               logging.debug("Generated PSBT: %s" % (psbt,))
+               sys.stderr.write("PSBT signing failed\n")
+               return None
+            block, signet_solution = decode_psbt(psbt_signed["psbt"])
```
Comment (ajtowns) [inline contrib/signet/miner:346]: Does it still make sense to grab the `block` from `decode_psbt` here? You already passed it in to `generate_psbt`. Could just be `_, signet_solution = decode_psbt(...)`, or perhaps have separate functions `get_solution_from_psbt()` and `get_block_from_psbt()` ?

**Review by ajtowns — COMMENTED** [review]
Concept ACK

**Comment by ajtowns** [comment]
> Using real mining equipment as opposed to a CPU miner is useful, because of the many quirks real devices have (e.g. version bit grinding). Afaik there's no good emulator.

FWIW, I did a patch for bitcoin-util to do version bit grinding: https://github.com/ajtowns/bitcoin/commit/634e72cbfc0aa3f657a35c7b597f688bb2bb29a6   (Note that it lightly depends on the changes in #28802)

File: test/functional/tool_signet_miner.py (hunk 1 above)
Comment (Sjors) [reply test/functional/tool_signet_miner.py:70]: Oops, should be `OP_TRUE`

File: contrib/signet/miner (hunk 5 above)
Comment (Sjors) [reply contrib/signet/miner]: I ended up making separate functions in 6f5419e3953781ade75d7986f293ce0fc13e90ce.

**Comment by Sjors** [comment]
Thanks for the review, I took the suggestions.

It could be useful to add version grinding support to OP_TRUE signets.

File: contrib/signet/miner [hunk 6]
```diff
@@ -97,10 +100,7 @@ def finish_block(block, signet_solution, grind_cmd):
         block.rehash()
     return block
 
-def generate_psbt(tmpl, reward_spk, *, blocktime=None, poolid=None):
-    signet_spk = tmpl["signet_challenge"]
-    signet_spk_bin = bytes.fromhex(signet_spk)
-
+def new_block(tmpl, reward_spk, blocktime=None, poolid=None):
```
Comment (ajtowns) [inline contrib/signet/miner:103]: Should keep the `, *,` to ensure `blocktime` and `poolid` are keyword-only args.

**Review by ajtowns — COMMENTED** [review]
ACK ecee62d2ab63685455b428db4e832827b6ce36f1 with nits

**Comment by Sjors** [comment]
Taken both suggestions.

I modified the test to actually check the signet commitment absence.

File: test/functional/tool_signet_miner.py [hunk 7]
```diff
@@ -26,26 +32,32 @@ def add_options(self, parser):
...
-        node = self.nodes[0]
-        # import private key needed for signing block
-        node.importprivkey(bytes_to_wif(CHALLENGE_PRIVATE_KEY))
+    def setup_network(self):
+        self.setup_nodes()
+        # Nodes with different signet networks are not connected
 
-        # generate block with signet miner tool
+    # generate block with signet miner tool
+    def mine_block(self, node):
```
Comment (ajtowns) [inline test/functional/tool_signet_miner.py:59]: Could consider checking manual mining as well:

```python
    # generate block with signet miner tool
    def mine_block_manual(self, node):
        assert_equal(node.getblockcount(), 1)
        base_dir = self.config["environment"]["SRCDIR"]
        signet_miner_path = os.path.join(base_dir, "contrib", "signet", "miner")
        base_cmd = [
                sys.executable,
                signet_miner_path,
                f'--cli={node.cli.binary} -datadir={node.cli.datadir}',
        ]

        template = node.getblocktemplate(dict(rules=["signet","segwit"]))
        genpsbt = subprocess.run(base_cmd + [
                'genpsbt',
                f'--address={node.getnewaddress()}',
                '--poolnum=98',
            ], check=True, input=json.dumps(template).encode('utf8'), capture_output=True)
        solvepsbt = subprocess.run(base_cmd + [
                'solvepsbt',
                f'--grind-cmd={self.options.bitcoinutil} grind',
            ], check=True, input=genpsbt.stdout, capture_output=True)
        node.submitblock(solvepsbt.stdout.decode('utf8').strip())
        assert_equal(node.getblockcount(), 2)

...
        node = self.nodes[2]
        self.log.info("Signet node with trivial challenge (push sha256 hash)")
        self.mine_block(node)
        assert(SIGNET_COMMITMENT not in get_segwit_commitment(node))
        self.mine_block_manual(node)
        assert(SIGNET_COMMITMENT not in get_segwit_commitment(node))
```

File: test/functional/tool_signet_miner.py [hunk 8]
```diff
@@ -61,6 +73,28 @@ def run_test(self):
...
+        self.log.info("Signet node with trivial challenge (OP_TRUE)")
+        self.mine_block(node)
+        # MAY omit signet commitment (BIP 325). Do so for better compatibility
+        # with signet unaware mining software and hardware.
+        assert(SIGNET_COMMITMENT not in get_segwit_commitment(node))
+
+        node = self.nodes[2]
+        self.log.info("Signet node with trivial challenge (push sha256 hash)")
+        self.mine_block(node)
+        assert(SIGNET_COMMITMENT not in get_segwit_commitment(node))
```
[the remark this answers is not in the document]
Comment (Sjors) [reply test/functional/tool_signet_miner.py]: Taken

File: contrib/signet/README.md [hunk 9]
```diff
@@ -44,6 +44,8 @@ Adding the --ongoing parameter will then cause the signet miner to create blocks
 
     $MINER --cli="$CLI" generate --grind-cmd="$GRIND" --address="$ADDR" --nbits=$NBITS --ongoing
 
+For custom signets with a trivial challenge a PSBT is not necessary. The miner detects this for `OP_TRUE`.
```
[the remark this answers is not in the document]
Comment (Sjors) [reply contrib/signet/README.md]: Moved the comment down and taken the suggestion.

File: test/functional/tool_signet_miner.py (hunk 7 above)
Comment (Sjors) [reply test/functional/tool_signet_miner.py:66]: Also taken, and added a signed example.

**Comment by Sjors** [comment]
Also rebased, just in case that helps with CI.

File: contrib/signet/README.md [hunk 10]
```diff
@@ -81,3 +81,4 @@ These steps can instead be done explicitly:
 
 This is intended to allow you to replace part of the pipeline for further experimentation (eg, to sign the block with a hardware wallet).
 
+For custom signets with a trivial challenge such as  `OP_TRUE` the walletprocesspsbt step can be skipped.
```
Comment (danielabrozzoni) [inline contrib/signet/README.md:84]: I would rephrase as "the walletprocesspsbt step *has* to be skipped", as it currently returns an error:

```
❯ $CLI -signet getblocktemplate '{"rules": ["signet","segwit"]}' | $MINER --cli="$CLI" genpsbt --address="$ADDR" | $CLI -signet -stdin walletprocesspsbt
error code: -22
error message:
Specified sighash value does not match value stored in PSBT	
```

**Review by danielabrozzoni — COMMENTED** [review]
Concept ACK

> However contrib/signet/miner can't handle this, as it fails with PSBT signing failed.

I can't seem to have contrib/signet/miner fail on master, am I doing something wrong?
I'm on 1d5b2406bb9ce619219a3b76608bd764a3b162c3,
I'm testing manually with `bitcoind -signet -signetchallenge=51` and:
```
❯ $MINER --cli="$CLI" generate --grind-cmd="$GRIND" --address="$ADDR" --nbits=$NBITS --ongoing
2024-09-13 18:29:59 INFO Mined block at height 24; next in -30h15m33s (mine)
2024-09-13 18:30:17 INFO Mined block at height 25; next in -30h10m28s (mine)
2024-09-13 18:30:20 INFO Mined block at height 26; next in -30h5m8s (mine)
2024-09-13 18:30:28 INFO Mined block at height 27; next in -29h59m53s (mine)
2024-09-13 18:30:32 INFO Mined block at height 28; next in -29h54m35s (mine)
2024-09-13 18:31:36 INFO Mined block at height 29; next in -29h50m16s (mine)
2024-09-13 18:31:41 INFO Mined block at height 30; next in -29h44m57s (mine)
2024-09-13 18:31:47 INFO Mined block at height 31; next in -29h39m40s (mine)
2024-09-13 18:32:02 INFO Mined block at height 32; next in -29h34m32s (mine)
```

File: contrib/signet/README.md (hunk 10 above)
Comment (Sjors) [reply contrib/signet/README.md:84]: > sighash value does not match

That's a bug I think, but I'm not sure what's causing it.

Comment (danielabrozzoni) [reply contrib/signet/README.md:84]: I think walletprocesspsbt is trying to sign the PSBT using `SIGHASH_DEFAULT`, while the PSBT specifies to use `SIGHASH_ALL`.

Using `$CLI -signet walletprocesspsbt $PSBT true ALL` works for me.

**Comment by Sjors** [comment]
> I can't seem to have contrib/signet/miner fail on master, am I doing something wrong?

It's possible that this changed in some recent rebase. Instead of failing it produces an empty witness and sticks that in the signet commitment.

I updated the PR description.

**Comment by Sjors** [comment]
Occasional rebase.

File: test/functional/tool_signet_miner.py [hunk 11]
```diff
@@ -60,7 +79,66 @@ def run_test(self):
...
+            signet_miner_path,
+            f'--cli={node.cli.binary} -datadir={node.cli.datadir}',
+        ]
+
+        template = node.getblocktemplate(dict(rules=["signet","segwit"]))
+        genpsbt = subprocess.run(base_cmd + [
+                'genpsbt',
+                f'--address={node.getnewaddress()}',
+                '--poolnum=98',
+            ], check=True, input=json.dumps(template).encode('utf8'), capture_output=True)
```
[the remark this answers is not in the document]
Comment (Sjors) [reply test/functional/tool_signet_miner.py:100]: I can reproduce this once I rebase on the latest master. I'll investigate.

[the remark this answers is not in the document]
Comment (Sjors) [reply test/functional/tool_signet_miner.py:100]: It seems to be caused by #31866, because the test passes when I revert it.

**Comment by Sjors** [comment]
Rebased after #31866 to use `--cli={shlex.join(rpc_argv)}` in `mine_block_manual`.

File: test/functional/tool_signet_miner.py [hunk 12]
```diff
@@ -12,14 +12,26 @@
...
 from test_framework.key import ECKey
-from test_framework.script_util import key_to_p2wpkh_script
+from test_framework.script_util import CScript, key_to_p2wpkh_script
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal
 from test_framework.wallet_util import bytes_to_wif
 
 
 CHALLENGE_PRIVATE_KEY = (42).to_bytes(32, 'big')
+SIGNET_COMMITMENT = 'ecc7daa2'
```
Comment (theStack) [inline test/functional/tool_signet_miner.py:23]: for consistency, would use the same name here as in the C++ codebase, i.e. `SIGNET_HEADER`
https://github.com/bitcoin/bitcoin/blob/3edf400b1020d7b88402ebc0e758b1fad2e7a781/src/signet.cpp#L27
could also move this to a test_framework module (maye `blocktools.py`) so it can be reused in the signet miner script

**Review by theStack — COMMENTED** [review]
Concept ACK

**Comment by Sjors** [comment]
Encoded it as bytes: `SIGNET_HEADER = b"\xec\xc7\xda\xa2"` and using the constant in `contrib/signet/miner`

File: test/functional/tool_signet_miner.py [hunk 13]
```diff
@@ -79,6 +80,38 @@ def mine_block(self, node):
             ], check=True, stderr=subprocess.STDOUT)
         assert_equal(node.getblockcount(), n_blocks + 1)
 
+    # generate block using the signet miner tool genpsbt and solvepsbt commands
+    def mine_block_manual(self, node, sign):
```
Comment (theStack) [inline test/functional/tool_signet_miner.py:84]: nit: could enforce `sign` as keyword argument in order to enhance readability at the call-sites below
```suggestion
    def mine_block_manual(self, node, *, sign):
```

File: contrib/signet/README.md [hunk 14]
```diff
@@ -80,3 +80,4 @@ These steps can instead be done explicitly:
...
```
Comment (theStack) [inline contrib/signet/README.md:83]: extra-space-nit:
```suggestion
For custom signets with a trivial challenge such as `OP_TRUE` the walletprocesspsbt step can be skipped.
```

**Review by theStack — APPROVED** [review]
Code-review ACK 6adec89e3961508875e79240a19bc981e5addf1a

Left a few non-blocking nits below. I was wondering whether OP_2...OP_16 should also be treated as trivial, i.e. extend the OP_TRUE check to a range (0x51-0x60)? On one hand I can't imagine why anyone would want to use those (and even if, they could always do it via the one byte longer `0102`...`0110` instead), on the other hand one could just include them for consistency reasons.

**Comment by Sjors** [comment]
> I was wondering whether OP_2...OP_16 should also be treated as trivial,

It may be useful to have multiple public signets out there and avoid them accidentally reorging each other.

**Comment by edilmedeiros** [comment]
> > I was wondering whether OP_2...OP_16 should also be treated as trivial,
> 
> It may be useful to have multiple public signets out there and avoid them accidentally reorging each other.

It would be useful indeed.

[BIP325](https://github.com/bitcoin/bips/blob/master/bip-0325.mediawiki) specifies a behavior that's impossible to detect in the general case, though:

> There is one other acceptable special case: if a block's challenge is e.g. `OP_TRUE` (`0x51`), where an empty solution would result in success, the block is also considered valid if the signet commitment is absent.

I think the current PR implements a good trade-off in the `trivial_challenge()` function (OP_TRUE and single by pushes). It can be expanded over time with more complex patterns if required.

**Comment by theStack** [comment]
@edilmedeiros: I agree it's fine as-is for now and can easily extended later. Note though that including OP_2...OP_16 is trivial, as it's just an extended range to OP_1/OP_TRUE, i.e. a diff like this would likely be sufficient:
```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
index 27f7afba8d..b79fa5d4f8 100755
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -245,8 +245,8 @@ def trivial_challenge(spkhex):
     challenges such as OP_TRUE
     """
     spk = bytes.fromhex(spkhex)
-    if len(spk) == 1 and spk[0] == 0x51:
-        # OP_TRUE
+    if len(spk) == 1 and 0x51 <= spk[0] <= 0x60:
+        # OP_TRUE/OP_1...OP_16
         return True
     elif 2 <= len(spk) <= 76 and spk[0] + 1 == len(spk):
         # Single fixed push of 1-75 bytes
```

File: contrib/signet/miner [hunk 15]
```diff
@@ -229,6 +238,21 @@ def seconds_to_hms(s):
...
 
+def trivial_challenge(spkhex):
+    """
+    BIP325 allows omitting the signet commitment when scriptSig and
+    scriptWitness are both empty. This is the case for trivial
+    challenges such as OP_TRUE
+    """
+    spk = bytes.fromhex(spkhex)
+    if len(spk) == 1 and spk[0] == 0x51:
+        # OP_TRUE
```
[the remark this answers is not in the document]
Comment (Sjors) [reply contrib/signet/miner:249]: I switched to using a range, where it makes more sense to use a code comment.

**Comment by edilmedeiros** [comment]
I agree with you, but I also prefer the current code; seems easier to understand.

**Comment by Sjors** [comment]
I added `OP_2` through `OP_16`, with a new test case for the latter. It's not practical to catch _all_ possible trivial scripts though.

**Review by theStack — APPROVED** [review]
re-ACK 2be93f7ba3866892d98795384779ce2e965753a1

**Review by theStack — APPROVED** [review]
re-ACK 6ee32aaaca4a46d42594bc16479868c76cc67fd8

(as per `$ git range-diff 2be93f7b...6ee32aaa`)

**Review by danielabrozzoni — COMMENTED** [review]
ACK 6ee32aaaca4a46d42594bc16479868c76cc67fd8

I'm not too familiar with the signet miner, but the code looks good to me. I ran the tests locally, and did some manual testing using `-signetchallenge=51` and checking that the signet commitment is absent.

**Comment by achow101** [comment]
ACK 6ee32aaaca4a46d42594bc16479868c76cc67fd8
